### PR TITLE
Changes to odd assumptions of Mul

### PR DESCRIPTION
**Pull request by debugger22**
@pelegm Would you like to review?

Fixes #8648

### CODE REVIEW

**Comment by debugger22** [comment]
@toolforger ping

File: sympy/core/mul.py [hunk 1]
```diff
@@ -1150,12 +1150,19 @@ def _eval_is_negative(self):
...
 
         if is_integer:
             r, acc = True, 1
+            _iter = -1
             for t in self.args:
+                _iter += 1
                 if not t.is_integer:
+                    if t.is_Rational:
+                        other = self.args[_iter ^ 1]
+                        if t.as_numer_denom()[1].is_even and sqrt(other).is_even:
```
Comment (debugger22) [inline sympy/core/mul.py]: I'm not confident about this `t.as_numer_denom()[1].is_even`(case: n=2, den=4) but somehow it's working.

File: sympy/core/mul.py [hunk 2]
```diff
@@ -1150,12 +1150,19 @@ def _eval_is_negative(self):
...
         is_integer = self.is_integer
 
         if is_integer:
             r, acc = True, 1
+            _iter = -1
             for t in self.args:
+                _iter += 1
                 if not t.is_integer:
+                    if t.is_Rational:
+                        other = self.args[_iter ^ 1]
```
Comment (asmeurer) [inline sympy/core/mul.py]: Can you add a comment explaining what this logic does?

File: sympy/core/mul.py [hunk 3]
```diff
@@ -1150,12 +1150,19 @@ def _eval_is_negative(self):
...
 
     def _eval_is_odd(self):
+        from sympy import sqrt
         is_integer = self.is_integer
 
         if is_integer:
             r, acc = True, 1
+            _iter = -1
             for t in self.args:
+                _iter += 1
```
Comment (asmeurer) [inline sympy/core/mul.py]: You can use `enumerate`

File: sympy/core/mul.py (hunk 2 above)
Comment (debugger22) [inline sympy/core/mul.py]: I'll do that. I just used it to flip between 0 and 1.

**Comment by asmeurer** [comment]
I don't understand what this fix is doing. It looks kind of off, given the strange `_iter ^ 1` bit, but maybe it is correct. 

Regardless if it's correct, it does seem a little specialized. For instance, `(Symbol('n', integer=True, even=True)**2/4).is_integer` still gives None with it (even though it's half `Symbol('n', integer=True, even=True)**2/2` which is now known to be even). Probably that means we should localize all complicated logic into `is_integer` and make `x.is_even` just check `x.is_integer and (x/2).is_integer` and `x.is_odd` check `(x + 1)`.is_even`. 

It also feels like we could somehow be using the diophantine solvers to do some heavy lifting here. I'm not entirely clear how, though.

File: sympy/core/mul.py (hunk 2 above)
Comment (asmeurer) [inline sympy/core/mul.py]:
```py
In [6]: [i ^ 1 for i in range(10)]
Out[6]: [1, 0, 3, 2, 5, 4, 7, 6, 9, 8]
```

Comment (asmeurer) [inline sympy/core/mul.py]: Remember that `&`, `|` and `^` work bitwise in Python.

**Comment by asmeurer** [comment]
@debugger22 feel free to ping me on assumptions pull requests. I may miss some (I do get behind on emails), but it gives it a higher chance that I will see it. And if I don't respond on something that you do want me to see, ping me on Gitter about it.

**Comment by debugger22** [comment]
There's a direct comparison with `S(2)` in the implementation of `is_integer`
```
    def _eval_is_integer(self):
        is_rational = self.is_rational

        if is_rational:
            n, d = self.as_numer_denom()
            print(n, d)
            if d is S.One:
                return True
            elif d is S(2):
                print("dA")
                return n.is_even
        elif is_rational is False:
            return False
```

**Comment by debugger22** [comment]
>  Probably that means we should localize all complicated logic into `is_integer` and make `x.is_even` just check `x.is_integer and (x/2).is_integer` and `x.is_odd` check `(x + 1)`.is_even`. 

This one looks good. As of now, they're all calling each others.

**Comment by toolforger** [comment]
Did you try the rewritten/added test cases from #9130?
They're currently on `@XFAIL`.

**Comment by debugger22** [comment]
Not yet. I'll pull that branch.
Whatever is proposed in this PR is probably not going to stay. I'm working on the implementation suggested by Aaron.

**Comment by debugger22** [comment]
@toolforger they still fail.

File: sympy/core/mul.py [hunk 4]
```diff
@@ -1153,23 +1153,28 @@ def _eval_is_negative(self):
...
 
     def _eval_is_odd(self):
+        from sympy import sqrt
         is_integer = self.is_integer
 
         if is_integer:
             r, acc = True, 1
-            for t in self.args:
-                if not t.is_integer:
+            for t1, t2 in enumerate(self.args):
```
Comment (asmeurer) [inline sympy/core/mul.py]: It's better to call the first item from enumerate `i`.

File: sympy/core/mul.py [hunk 5]
```diff
@@ -1153,23 +1153,28 @@ def _eval_is_negative(self):
...
+                if not t2.is_integer:
+                    if t2.is_Rational:
```
Comment (asmeurer) [inline sympy/core/mul.py]: I believe a rational term in a Mul will always be the first arg.

**Comment by toolforger** [comment]
> @toolforger they still fail.

Ah well, pity - it would have been nice to un-XFAIL them, but if that 
wasn't the focus of your work, then so be it.

File: sympy/core/mul.py [hunk 6]
```diff
@@ -1153,23 +1153,28 @@ def _eval_is_negative(self):
...
+                        other = self.args[t1 ^ 1]  # get 0 if 1 or 1 if 0
```
Comment (asmeurer) [inline sympy/core/mul.py]: I'm still concerned that this still arbitrarily only looks at the first or second argument of the Mul. Remember that Mul can have `n` args.

Comment (asmeurer) [inline sympy/core/mul.py]: Actually, because of [this](https://github.com/sympy/sympy/pull/9258/files#r30987928), `t1` will always be 0.

Comment (asmeurer) [inline sympy/core/mul.py]: For example

```py
In [11]: x = Symbol('x', integer=True)

In [12]: y = Symbol('y', even=True)

In [13]: Mul(S.Half, y**2, x).is_even

In [14]: Mul(S.Half, y**2, x, evaluate=False).is_even
Out[14]: True

In [16]: Mul(S.Half, y**2, x, evaluate=False).args
Out[16]:
⎛      2   ⎞
⎝1/2, y , x⎠

In [17]: Mul(S.Half, y**2, x).args
Out[17]:
⎛         2⎞
⎝1/2, x, y ⎠
```

File: sympy/core/mul.py (hunk 5 above)
Comment (Upabjojr) [inline sympy/core/mul.py]:
```
In [1]: m = Mul(x, 3, y, evaluate=False)

In [2]: m.args
Out[2]: (x, 3, y)
```

Comment (asmeurer) [inline sympy/core/mul.py]: That will likely break things, though. As far as I know, Mul assumes this in a lot of places. @smichr

Comment (Upabjojr) [inline sympy/core/mul.py]: Maybe one should use _expr.as_coeff_mul()_ in such cases?

Comment (asmeurer) [inline sympy/core/mul.py]: Yes, I think we should use that function (or `as_coeff_Mul`, depending on whether you want the args or not).

**Comment by jksuom** [comment]
It might be practical to base ``Mul.is_even/odd`` on a common ``Mul.parity``.

It would start with a check of ``self.is_rational``; if not, return False or None.
and then set ``n, d = self.as_numer_denom()``.

If ``d`` is not 1 or 2 (or possibly a power of 2), return None.
Then count the factors of ``n`` (which should be a Mul) and divide them into four classes.
* Even integers (``e``)
* Odd integers
* Integers of unknown parity (``u``)
* Other factors

If one of last class is encountered, return None. 
Odd factors can be ignored.
Even factors of the type ``b**m`` should be counted with multiplicity (adding ``m`` to ``e``).
```
if d == 1:
    if e > 0:
        return 0   # even parity regardless of factors of unknown parity
    elif u == 0: 
        return 1   # odd parity, odd factors only
    else:
        return None  # unknown parity
```
Less can be said in the other case.
```
if d == 2:
    if e > 1:
        return 0    # one even factor is divided by 2 but another remains
    else:
        return None
```
``Mul.is_even/odd`` should call ``Mul.parity`` and translate the returned 0 or 1 as needed. None and False are passed over as such.

This last piece of code could be extended to the case where ``d`` is a power of 2: ``d == 2**k``.
The test would then become ``if e > k:``.

**Comment by certik** [comment]
Looks like there is some feedback to be addressed by @debugger22.

**Comment by debugger22** [comment]
Yes. I'm sorry for the delay. I'll work on it very soon.

**Comment by asmeurer** [comment]
@jksuom that sounds like a good start. I think you can also extend it to the case where `n` is an `Add` by checking the parity of each term (handling `Add` is important if we want to be able to reduce `x.is_odd` to `(x + 1).is_even` or some similar).

**Comment by asmeurer** [comment]
@debugger22 I don't like this pull request at all as it is. If you want to do more, feel free to reopen, but I am closing it. A better PR would probably have to be completely rewritten anyway.

**Pull request closed without merging**